### ISSUE TITLE
Using Gain and Qini for cross validation

### ISSUE THREAD

**Issue by shaddyab** [uber/causalml#279]
I have a question regarding the [meta_learners_with_synthetic_data] example (https://github.com/uber/causalml/blob/master/examples/meta_learners_with_synthetic_data.ipynb).

When tuning the meta-learners via cross validation, 
1. should the cum gain or qini auuc be used as the metric to select the appropriate hyperparameters or should other regression metrics be used?
2. In the case of gain/qini should the area under the random curve be subtracted? 
3. Are these metrics rigid enough to handle that the ATE in each training or validation fold may be slightly different (hence we are not comparing apples-to-apples)?

**Comment by vferraz** [comment]
I am also facing similar problems, especially because these metrics can be challenging for optimization models. 

I am running hyperparameter tuning for an uplift random forest classifier, and I get AUUC scores anywhere between 0.2 and 20 something. It might well be due to overfitting, but I don't understand how absurdly divergent values can be produced by fitting the same model to the same dataset and applying the same calculation functions. 

Does anyone know any best practices for hyperparameter tuning in such cases? Especially because this can be performance costly and every failed trial takes hours to be completed.

**Comment by JDCBean** [comment]
> I am also facing similar problems, especially because these metrics can be challenging for optimization models.
> 
> I am running hyperparameter tuning for an uplift random forest classifier, and I get AUUC scores anywhere between 0.2 and 20 something. It might well be due to overfitting, but I don't understand how absurdly divergent values can be produced by fitting the same model to the same dataset and applying the same calculation functions.
> 
> Does anyone know any best practices for hyperparameter tuning in such cases? Especially because this can be performance costly and every failed trial takes hours to be completed.

I also am having these same issues. When I have experimental data (randomized) with four treatment arms and one control group, when I compare each treatment to the control, I have some wildly divergent values for the Qini score - sometimes just for one treatment-control comparison. Also, shouldn't the random Qini curve be at least somewhat resembling any random curve with and approximate diagonal to represent the 50/50 randomness.

I'm confused as to how this seems so erratic at times. Any help is definitely appreciated in understanding this more. Thanks.

**Comment by jeongyoonlee** [comment]
The Qini and AUUC scores use the cumulative outcome difference between treatment and control. However, they can be unstable due to an imbalanced treatment distribution, a small sample size, and an overfitted treatment estimation. 

We have the option to use TMLE with cross-validation to address the instability of the Qini and AUUC scores/plots as follows:
* [`plot(..., tmle=True, ...)`](https://github.com/uber/causalml/blob/86e5c89321a62384dd7d0eca61b8d96ecafa4892/causalml/metrics/visualize.py#L20)
* [`auuc_score(..., tmle=True, ...)`](https://github.com/uber/causalml/blob/86e5c89321a62384dd7d0eca61b8d96ecafa4892/causalml/metrics/visualize.py#L786)
* [`qini_score(..., tmle=True, ...)`](https://github.com/uber/causalml/blob/86e5c89321a62384dd7d0eca61b8d96ecafa4892/causalml/metrics/visualize.py#L820) 
* [Example notebook](https://github.com/uber/causalml/blob/master/docs/examples/validation_with_tmle.ipynb)

Just so you know, it's computationally more expensive than the vanilla Qini and AUUC, but if you have a small sample size and/or skewed treatment distribution, it might be worth trying.

I hope this helps.
